### PR TITLE
Add Elixir libs.

### PR DESCRIPTION
**Pull request by cpursley**
This assumes [Plug](https://github.com/elixir-plug/plug) is being used.

Somewhat follows the Ruby library but passes secret as an arg (to allow for multiple, dynamic signing secrets).

Review from folks who know Elixir very welcome!

### CODE REVIEW

**Comment by tasn** [comment]
Thanks a lot!

@mmcc, any thoughts?

**Review by mmcc — APPROVED** [review]
This looks good to me, thank you! Only real thoughts are on the README.

- Do we want to mention Webhoox there? It's a neat/complimentary project, and we call it out in the codebase anyway.
- We're missing the actual package name in the readme. I assume that's because we haven't published it yet, but should we go ahead and include a placeholder or other installation instructions?

**Comment by cpursley** [comment]
I don't believe Webhoox is mentioned in the README, but there is a mention [here](https://github.com/standard-webhooks/standard-webhooks/pull/87/files#diff-eae091c3ca2864cb1c92954076c93d57fc8448c38ba1fe7bf56bb70737fc9702R4) that I'd be glad to remove.

Would you like to publish an actual [hex](https://hex.pm) package? If so, I can whip up package name stuff.

**Comment by tasn** [comment]
Just to add on what @mmcc said: maybe have the scaffolding for the hex package (assuming that's the packaging in elixir land), and maybe add: (1) a release CI task, (2) a linting CI task?

**Comment by zekth** [comment]
Currently our release process is not clearly defined, however please add a `lint/test` ci workflow. You can look at https://github.com/standard-webhooks/standard-webhooks/blob/main/.github/workflows/go-lint.yml for example

**Comment by mmcc** [comment]
> I don't believe Webhoox is mentioned in the README, but there is a mention [here](https://github.com/standard-webhooks/standard-webhooks/pull/87/files#diff-eae091c3ca2864cb1c92954076c93d57fc8448c38ba1fe7bf56bb70737fc9702R4) that I'd be glad to remove.

Sorry I meant the opposite! The project seems complimentary, feels worth noting in the README as an option.

For linting, we might want to go further eventually, but for now should at least just enforce people are running `mix format`. Using `mix format --check-formatted` should be enough in CI.

**Comment by cpursley** [comment]
That all sounds good. And we could actually use this (Standard Webhooks) lib in the walex project going forward once there's a package on hex. I'll work something up soon, thank you for the feedback.

File: libraries/elixir/mix.exs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+defmodule StandardWebhooks.MixProject do
```
Comment (themusicman) [inline libraries/elixir/mix.exs:1]: It is missing some of the required fields to publish the package to hex. https://hex.pm/docs/publish#example-mixexs-file

File: .github/workflows/elixir-lint.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Run tests
+        run: |
+          mix test
+        working-directory: libraries/elixir
+
+        # See mix.exs aliases quality task
+      - name: Check code quality, security and format
+        run: |
+          mix quality --strict
```
Comment (themusicman) [inline .github/workflows/elixir-lint.yml:52]: The `quality` alias runs the `mix format --check-formatted` a second time.

**Comment by cpursley** [comment]
I think this should do it. I didn't really have a good way to test if the workflows run correctly (I think y'all need to approve the workflows to run).

A hex account and `HEX_API_KEY` will need to be set up for the releases: https://hex.pm/docs/publish#publishing-from-ci

Let me know what else it needs.

**Review by zekth — APPROVED** [review]
:v:

File: libraries/elixir/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+Elixir library for Standard Webhooks
+
+Need to handle inbound webhooks in Elixir? Check out [webhoox](https://github.com/cpursley/webhoox).
```
Comment (tasn) [inline libraries/elixir/README.md:3]: I think this reads too much like an ad. Is there a more subtle way of doing it, e.g. "this library was based on code from [webhoox]..."?

File: libraries/elixir/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+# Example
+
+Verifying a webhook payload:
+
+```elixir
+StandardWebhooks.verify(plug_conn, webhook_payload, webhook_secret)
```
Comment (tasn) [inline libraries/elixir/README.md:10]: Sorry for coming late in the party, but I'm not a fan of it having a different API to the rest of the libs, I think there's value in having all the libs be very similar.

On a personal level I also prefer the other way in general. Creating the other structure should be essentially zero cost, and it centralizes the parsing of the secret to one place so you know whether your secret is malformed before it's time to verify, but also you get to reuse this parsed structure.

Though it's really more about the APIs being the same across all of the libs.

File: libraries/elixir/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+Elixir library for Standard Webhooks
+
+# Example
+
+Verifying a webhook payload:
+
+```elixir
+StandardWebhooks.verify(webhook_payload, plug_conn, webhook_secret)
```
Comment (tasn) [inline libraries/elixir/README.md:8]: @cpursley, not just the ordering, but also having an instance.

Here is how it looks like in Ruby (which you mentioned):
```ruby
wh = StandardWebhooks::Webhook.new(base64_secret)
wh.verify(webhook_payload, webhook_headers)
```

Is it possible to do it here too, or is it very much not how people do it in Elixir? That's what I was referring to in https://github.com/standard-webhooks/standard-webhooks/pull/87/#discussion_r1438942237

Comment (themusicman) [reply libraries/elixir/README.md:8]: Elixir doesn't have instances. It is a functional language without objects. It has structs like golang but you can't bind methods them.

Comment (tasn) [reply libraries/elixir/README.md:8]: Gotcha, sorry for the noise. :)

**Review by tasn — APPROVED** [review]
Approving based on the previous approves (changes were minor).

**Comment by tasn** [comment]
Sure, happy to elaborate!

I was referring to `StandardWebhooks.verify(webhook_payload, plug_conn, webhook_secret)`. This function currently does two things: (1) parses/validates the secret, and (2) verifies the payload. This means that the parsing/validating of the secret is done only when we try to verify and not at the time initialization. This defers an error (which is better caught early), but also essentially means the verify function does extra things that are not really "verify". That's why we did it that way in the other libs, have one step for parsing the param, and another for actually doing the verification. This also makes it so verifies can reuse that already parsed secret (though admittedly that's an insignificant optimization, at least with the current implementations).

**Pull request merged**